v1/amqp/amqp.go

```diff
@@ -15,39 +15,58 @@
 package amqp
 
 import (
+	"context"
 	"sync"
+	"time"
 
 	"github.com/Azure/go-amqp"
 
 	cloudevents "github.com/cloudevents/sdk-go/v2"
 	"github.com/redhat-cne/sdk-go/pkg/channel"
 	"github.com/redhat-cne/sdk-go/pkg/errorhandler"
 	amqp1 "github.com/redhat-cne/sdk-go/pkg/protocol/amqp"
+	log "github.com/sirupsen/logrus"
 )
 
 var (
-	instance *AMQP
-	once     sync.Once
+	instance      *AMQP
+	retryTimeout  = 500 * time.Millisecond
+	cancelTimeout = 30 * time.Second
 )
 
-//AMQP exposes amqp api methods
+// AMQP exposes amqp api methods
 type AMQP struct {
 	Router *amqp1.Router
 }
 
-//GetAMQPInstance get event instance
+// GetAMQPInstance get event instance
 func GetAMQPInstance(amqpHost string, dataIn <-chan *channel.DataChan, dataOut chan<- *channel.DataChan, closeCh <-chan struct{}) (*AMQP, error) {
-	once.Do(func() {
-		router, err := amqp1.InitServer(amqpHost, dataIn, dataOut, closeCh)
-		if err == nil {
-			instance = &AMQP{
-				Router: router,
-			}
+	ctx, cancel := context.WithTimeout(context.Background(), cancelTimeout)
+	defer cancel()
+	var router *amqp1.Router
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, errorhandler.AMQPConnectionError{}
+		default:
 		}
-	})
+		router, err = amqp1.InitServer(amqpHost, dataIn, dataOut, closeCh)
+		if err != nil {
+			log.Info("retrying connecting to amqp.")
+			time.Sleep(retryTimeout)
+			continue
+		}
+
+		instance = &AMQP{
+			Router: router,
+		}
+		break
+	}
 	if instance == nil || instance.Router == nil {
-		return nil, errorhandler.AMQPConnectionError{Desc: "amqp connection error"}
+		return nil, errorhandler.AMQPConnectionError{}
```

Comment:
why removed description ?

Reply:
text might be used for parsing some where

Reply:
The `AMQPConnectionError` itself already has description.
Before removing description:

> ERRO[0000] error starting amqp at amqp:localhost:5672 error: amqp connection error amqp connection error 
> WARN[0000] requires QPID router installed to function fully amqp connection error amqp connection error 

After removing description:

> ERRO[0000] error starting amqp at amqp:localhost:5672 error: amqp connection error  
> WARN[0000] requires QPID router installed to function fully amqp connection error

Reply:
nice